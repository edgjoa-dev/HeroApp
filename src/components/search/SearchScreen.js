import React, { useMemo } from 'react'
import {useNavigate, useLocation} from 'react-router-dom'
import queryString from 'query-string'
import { useForm } from '../../Hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../heros/HeroCard';

export const SearchScreen = () => {

    const navigate = useNavigate();
    const location = useLocation();

    const {q = ''} = queryString.parse(location.search);

    const [FormValues, handleInputChange] = useForm({
        searchText: q,
    });

    const {searchText} = FormValues;

    const heroesFileted = useMemo(() => getHeroesByName(q), [q]) 
    
    



    const handleSearch = (e) => {   
        console.log(searchText);
        e.preventDefault();
        navigate(`?q=${searchText}`)
    };
    


    return (
        <div>
            <h1>Search</h1>
            <hr/>

            <div className="row" >

                <div className="col-5" >

                    <form  onSubmit={handleSearch}
                        className="m-3" >
                        <input 
                        type="text" 
                        placeholder="Search a Hero"
                        name='searchText'
                        autoComplete="off"
                        value={searchText}
                        onChange={handleInputChange}
                        className="row"
                        />

                        <button 
                        className="row btn btn-outline-primary mt-1 "
                        type="submit"
                        >
                            Buscar                        
                        </button>

                    </form>
                </div>

                <div className="col-7" >

                <h4>Resultados</h4> 
                <hr/>

                    {
                        (q === '')
                        ? <div className="alert alert-info">Buscar un HÃ©roe</div>
                        : (heroesFileted.length === 0) && <div className="alert alert-danger">No se encontraron resultados de: {q}</div>
                    }

                    {
                        

                        heroesFileted.map( hero => (
                            <HeroCard 
                            key={hero.id}
                            {...hero}
                            />
                        ))

                    }
                </div>
            </div>
        </div>
    )
}
