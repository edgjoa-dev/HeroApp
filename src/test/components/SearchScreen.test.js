import {mount} from 'enzyme'
import { MemoryRouter } from 'react-router-dom';
import { SearchScreen } from '../../components/search/SearchScreen';


const mockNavigate = jest.fn();


jest.mock('react-router-dom', ()=> ({

    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate,
}))


describe('Prueba en componente Search', () => {

    test('Debe de mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen /> 
            </MemoryRouter>
            );
        
            expect( wrapper).toMatchSnapshot();
            expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un HÃ©roe');


    });

    test('Debe de mostrar a Batman y el input con el valor del queryString', () => {
                
        const wrapper = mount( 
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen /> 
            </MemoryRouter>
            );

            expect(wrapper).toMatchSnapshot();
            expect(wrapper.find('input').prop('value')).toBe('batman');


    });


    test('Debe de mostrar alerta de error si no se encuentra la entrada del input de busqueda de heroes', () => {
                
        const wrapper = mount( 
            <MemoryRouter initialEntries={['search?q=batman123']}>
                <SearchScreen /> 
            </MemoryRouter>
            );

            expect(wrapper.find('.alert-danger').text().trim()).toBe('No se encontraron resultados de: batman123');
    });


    test('Debe llamar al navigate a la nueva pantalla', () => {
        
        const wrapper = mount( 
            <MemoryRouter initialEntries={['search']}>
                <SearchScreen /> 
            </MemoryRouter>
            );

            wrapper.find('input').simulate('change', {
                target: {
                    name: 'searchText',
                    value: 'batman'
                }
            })

            wrapper.find('form').prop('onSubmit')({
                preventDefault: () => {}
            })

            expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');
    });
    
    
    
})
