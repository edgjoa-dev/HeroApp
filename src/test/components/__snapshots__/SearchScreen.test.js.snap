// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en componente Search Debe de mostrar a Batman y el input con el valor del queryString 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        className="m-3"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="row"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value="batman"
        />
        <button
          className="row btn btn-outline-primary mt-1 "
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                className="text-info"
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                 Más... 
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Prueba en componente Search Debe de mostrar alerta de error di no se encuentra la entrada del input de busqueda de heroes 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        className="m-3"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="row"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value="123"
        />
        <button
          className="row btn btn-outline-primary mt-1 "
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No se encontraron resultados de: 
        123
      </div>
    </div>
  </div>
</div>
`;

exports[`Prueba en componente Search Debe de mostrar alerta de error si no se encuentra la entrada del input de busqueda de heroes 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        className="m-3"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="row"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value="123"
        />
        <button
          className="row btn btn-outline-primary mt-1 "
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No se encontraron resultados de: 
        123
      </div>
    </div>
  </div>
</div>
`;

exports[`Prueba en componente Search Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <form
        className="m-3"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="row"
          name="searchText"
          onChange={[Function]}
          placeholder="Search a Hero"
          type="text"
          value=""
        />
        <button
          className="row btn btn-outline-primary mt-1 "
          type="submit"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un Héroe
      </div>
    </div>
  </div>
</div>
`;
